java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -0,0 +1,290 @@
+      if (owner.isMethodSymbol() && ((JavaSymbol.MethodJavaSymbol) owner).typeVariableTypes.contains(argType)) {
+        typeVariablesToInfer.add((JavaType.TypeVariableJavaType) argType);
+      }
+    }
+
+    if (!typeVariablesToInfer.isEmpty()) {
+      for (int i = 0; i < formals.size(); i++) {
+        JavaType formalType = formals.get(i);
+        JavaType argType = argTypes.get(i);
+        boolean isLastParam = i == formals.size() - 1;
+
+        substitution = inferTypeSubstitution(method, substitution, isLastParam, formalType, argType);
+
+        if (substitution.typeVariables().containsAll(method.typeVariableTypes)) {
+          // we found all the substitution
+          break;
+        }
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, boolean isLastParam, JavaType formalType,
+    JavaType argType) {
+    if (formalType.isTagged(JavaType.TYPEVAR)) {
+      completeSubstitution(currentSubstitution, formalType, argType);
+    } else if (formalType.isArray() && argType.isArray()) {
+      completeSubstitution(currentSubstitution, ((JavaType.ArrayJavaType) formalType).elementType, ((JavaType.ArrayJavaType) argType).elementType);
+    } else if (formalType.isArray() && method.isVarArgs() && isLastParam) {
+      completeSubstitution(currentSubstitution, ((JavaType.ArrayJavaType) formalType).elementType, argType);
+    } else if (isParametrizedType(formalType) && isParametrizedType(argType)) {
+      List<JavaType> formalTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+      List<JavaType> argTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+      return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+    } else if (isParametrizedType(formalType) && !isParametrizedType(argType)) {
+      if (formalType.erasure() == argType || argType == symbols.nullType) {
+        List<JavaType> formalTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+        List<JavaType> fakeTypes = new ArrayList<>(formalTypeSubstitutedTypes.size());
+        for (int j = 0; j < formalTypeSubstitutedTypes.size(); j++) {
+          fakeTypes.add(symbols.objectType);
+        }
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, fakeTypes);
+        return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+      } else if (argType.isSubtypeOf(formalType.erasure()) && argType.isClass()) {
+        for (JavaType superType : ((JavaType.ClassJavaType) argType).symbol.superTypes()) {
+          if (formalType.erasure() == superType.erasure()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)